crates/core/src/views.rs

```diff
@@ -206,76 +260,98 @@ fn powersync_trigger_update_sql_impl(
 ) -> Result<String, SQLiteError> {
     let table = args[0].text();
 
-    let statement = extract_table_info(ctx.db_handle(), table)?;
+    let table_info = TableInfo::parse_from(ctx.db_handle(), table)?;
 
-    let name = statement.column_text(0)?;
-    let view_name = statement.column_text(1)?;
-    let local_only = statement.column_int(2)? != 0;
-    let insert_only = statement.column_int(3)? != 0;
-    // TODO: allow accepting a column list
-    let include_old = statement.column_type(4)? == sqlite::ColumnType::Text;
-    let include_metadata = statement.column_int(5)? != 0;
+    let name = &table_info.name;
+    let view_name = &table_info.view_name;
+    let insert_only = table_info.flags.insert_only();
+    let local_only = table_info.flags.local_only();
 
     let quoted_name = quote_identifier(view_name);
     let internal_name = quote_internal_name(name, local_only);
     let trigger_name = quote_identifier_prefixed("ps_view_update_", view_name);
     let type_string = quote_string(name);
 
     let db = ctx.db_handle();
-    let stmt2 = db.prepare_v2("select json_extract(e.value, '$.name') as name from json_each(json_extract(?, '$.columns')) e").into_db_result(db)?;
-    stmt2.bind_text(1, table, sqlite::Destructor::STATIC)?;
-    let json_fragment_new = json_object_fragment("NEW", &stmt2)?;
-    stmt2.reset()?;
-    let json_fragment_old = json_object_fragment("OLD", &stmt2)?;
-    let old_fragment: String;
-    let metadata_fragment: &str;
-
-    if include_old {
-        old_fragment = format!(", 'old', {:}", json_fragment_old);
-    } else {
-        old_fragment = String::from("");
+    let mut columns = ColumnNameAndTypeStatement::new(db, table)?;
+    let json_fragment_new = json_object_fragment("NEW", &mut columns.names_iter())?;
+    let json_fragment_old = json_object_fragment("OLD", &mut columns.names_iter())?;
+
+    let mut old_values_fragment = match &table_info.diff_include_old {
+        None => None,
+        Some(DiffIncludeOld::ForAllColumns) => Some(json_fragment_old.clone()),
+        Some(DiffIncludeOld::OnlyForColumns { columns }) => {
+            let mut iterator = columns.iter();
+            let mut columns =
+                streaming_iterator::from_fn(|| -> Option<Result<&str, ResultCode>> {
+                    Some(Ok(iterator.next()?.as_str()))
+                });
+
+            Some(json_object_fragment("OLD", &mut columns)?)
+        }
+    };
+
+    if table_info.flags.include_old_only_if_changed() {
+        // We're setting ignore_removed_columns here because values that are present in the new
+        // values but not in the old values should not lead to a null entry.
```

Comment:
Can you explain the reasoning here - why it should not lead to a null entry?

It would also be good to have a unit test for this specific case, demonstrating the behavior.

From my own testing, this appears to specifically be applicable to a case where a value is changed from null to another value. With `ignore_removed_columns`, `old.<column>` would not be present. Without it, it would be null. Either way you know it has changed due to the column being present in the "new" values. So I think it should be fine either way, but I'm still wondering on the specific reasoning here.

Reply:
I think I ran into this writing the test `crud tests tracks previous values for updates combined column filter and only tracking changes`. The problem is that we're passing `json_fragment_new` to `powersync_diff` - which includes all columns. The `old_values_fragment` however, does not. So there was a possibility where unchanged columns would be reported just because they're absent in `old_values_fragment` as they're not matched by the filter.

I don't remember why I came up with that workaround - a better solution is to just apply the same filter to the fragment producing the new columns too. That's what I've done now.